lib/utils.js

```diff
@@ -606,6 +606,77 @@ function decorateWithReadConcern(command, coll) {
   }
 }
 
+const emitProcessWarning = msg => process.emitWarning(msg, 'DeprecationWarning');
+const emitConsoleWarning = msg => console.error(msg);
+const emitDeprecationWarning = process.emitWarning ? emitProcessWarning : emitConsoleWarning;
+
+/**
+ * Default message handler for generating deprecation warnings.
+ *
+ * @param {string} [name] function name
+ * @param {string} [option] option name
+ * @return {string} warning message
+ * @ignore
+ * @api private
+ */
+function defaultMsgHandler(name, option) {
+  return `${name} option [${option}] is deprecated and will be removed in a later version.`;
+}
+
+/**
+ * Deprecates a given function's options.
+ *
+ * @param {object} [config] configuration for deprecation
+ * @param {string} [config.name] function name
+ * @param {Array} [config.deprecatedOptions] options to deprecate
+ * @param {number} [config.optionsIndex] index of options object in function arguments array
+ * @param {function} [config.msgHandler] optional custom message handler to generate warnings
+ * @param {function} [fn] the target function of deprecation
+ * @return {function} modified function that warns once per deprecated option, and executes original function
+ * @ignore
+ * @api private
+ */
+function deprecateOptions(config, fn) {
+  if (process.noDeprecation === true) {
+    return fn;
+  }
+
+  const msgHandler = config.msgHandler ? config.msgHandler : defaultMsgHandler;
+
+  const optionsWarned = new Set();
+  function deprecated() {
+    const options = shallowClone(arguments[config.optionsIndex]);
```

Comment:
why are we shallow-cloning this again? I feel like I asked for this at some point, but I can't remember why...

Reply:
Also, you should probably verify that `options` is an object, and only run your loop if it is.

Reply:
For some reason, when we don't shallow clone the options, a bunch of tests break (even though we don't think we are mutating the options object...)

Reply:
^ nevermind - the other tests failed without `shallowClone` because `shallowClone` returns an object no matter what it's called on, so when `optionsIndex` wasn't pointing to an object in other tests, it still assigned `options` to a valid (empty) object. Without `shallowClone`, there were errors in other tests where the deprecate function attempted to access properties of `undefined`.

 If we remove `shallowClone` and instead just check that `options` is an object, all the tests work.